src/zope/file/download.py

```diff
@@ -56,8 +56,12 @@ def getHeaders(context, contentType=None, downloadName=None,
         downloadName = downloadName or context.__name__
         if contentDisposition:
             if downloadName:
+                # Headers must be native strings. Under Py2, we probably
+                # need to encode the name
+                if str is bytes and not isinstance(downloadName, bytes):
+                    downloadName = downloadName.encode("utf-8")
                 contentDisposition += (
-                    '; filename="%s"' % downloadName.encode("utf-8")
+                    '; filename="%s"' % downloadName
```

Comment:
In WSGI, all headers must be native strings. In the case of Python 3, it's [up to the WSGI server](https://github.com/benoitc/gunicorn/issues/1202) to encode them for transmission on the wire, almost always with `latin-1`. Latin-1 can encode the first 256 unicode code points, handling many common characters like `\xe0` (à), but nothing beyond that.

In this code, Python 2 is still encoding all unicode strings using utf-8 (just as it used to do). Python 3 is leaving such strings alone and passing them to the WSGI server (as they must be `str`, not encoded bytes).

In this scenario, a string like `downloadName = u'Pile of \U0001F4A9'` will be handled as expected in Python 2, but in Python 3 is liable to produce a UnicodeEncodeError, depending on the server.

We *could* try to smuggle characters greater than 256 under Python 3 using `downloadName.encode('utf-8').decode('latin-1')`, thus passing a native str to the WSGI server (`'pile ð\x9f\x92©'`), which encodes in latin-1 to `b'pile \xf0\x9f\x92\xa9'`, and can be *decoded* using utf-8 to get the original string. Decoding in latin-1 produces gibberish, of course. How that's handled is up to the client. If the Python 2 code is correct in assuming that the client will decode using UTF-8, then that works out fine.